src/i2c.rs

```diff
@@ -207,118 +207,137 @@ where
     }
 }
 
-impl<I2C, SDA, SCL> embedded_hal::blocking::i2c::Read for Master<I2C, SDA, SCL>
+impl<I2C, SDA, SCL> embedded_hal::i2c::blocking::Read for Master<I2C, SDA, SCL>
 where
     I2C: I2c,
     SDA: OutputPin + InputPin,
     SCL: OutputPin,
 {
-    type Error = EspError;
+    type Error = embedded_hal::i2c::ErrorKind;
```

Comment:
OK so after reading [this](https://github.com/rust-embedded/embedded-hal/issues/229) I definitely believe we *should* change the error handling in this PR before it gest merged.

The main theme of the V1.0 error handling is to (a) allow the driver implementations to report "standardized" errors (a.k.a. `xyz::ErrorKind`) so that the driver-utilization code is driver impl agnostic, yet at the same time (b) **preserve** the original error (which is the `esp-idf-sys::EspError` for our case) in case somebody wants to analyze that one further and/or report it for diagnostics.

I think the current error handling in that PR achieves neither, as (a) almost always maps all `EspError` instances to just `ErrorKind::Other` without doing any sort of analysis on the actual EspError. Now, that's an OK initial step as long as (b) the error handling code preserved the original `EspError`, but it doesn't. We should fix that.

Or to put it another way, we should invent something more complex than
```rust
type Error = embedded_hal::i2c::ErrorKind;
```

Here's my proposal how this can be achieved (I'll do it in the context of the I2C driver, but it should be applicable for all other drivers as well):

So in the I2c trait impl, we define the `Error` associated type to be:
```rust
type Error = I2cError;
```

And define `I2cError` as follows:
```rust
#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub struct I2cError {
	kind: embedded_hal::i2c::ErrorKind,
	cause: EspError,
}

impl I2cError {
    pub fn new(kind: embedded_hal::i2c::ErrorKind, cause: EspError) -> Self {
		Self {
			kind,
			cause,
		}
	}
	
	pub fn other(cause: EspError) -> Self {
		Self::new(embedded_hal::i2c::ErrorKind::Other, cause)
	}
}

impl From<EspError> for I2cError {
    fn from(e: EspError) -> Self {
		Self::other(e)
	}
}

impl embedded_hal::i2c::Error for I2cError {
    fn kind(&self) -> embedded_hal::i2c::ErrorKind {
        self.kind
    }
}

impl Display for I2cError {
	// Call .display() on the `kind` member as well as .display() on the `cause` member
}

/// Extra ergonomics in STD mode, better compatibility with the anyhow crate
#[cfg(feature = "std")]
impl std::error::Error for I2cError {}
```

(By the way, we can also implement a macro - say - `hal-error!(I2cError, embedded_hal::i2c::ErrorKind)` that defines such an error type by a passed type name and the type name of the concrete trait-specific `ErrorKind`, so that we don't have to repeat the above verbosity for all drivers.)

So... the idea would be:
* If you don't want to analyze the original `EspError`, you just do `...map_err(Into::into)` and this gives you an I2cError instance with `kind = ErrorKind::Other` and `cause = <the original ESP-IDF error>`
* If you do want to analyze the original `EspError` to possibly return a more meaningful `ErrorKind`, you'll do:
```rust
...map_err(|e| I2cError::new(
       match e.code() {
	       ESP_ERR_TIMEOUT => ErrorKind::Whatever,
           ...,
		   _ => ErrorKind::Other,
	   },
	   cause,
   ))
```

Reply:
Going just with `embedded_hal::i2c::ErrorKind::Other` was the quickest way to covert to new HAL, but indeed we lost original error during this. I am looking into it today.